sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala

```diff
@@ -399,6 +395,52 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
   }
 
   /**
+   * Loads an `Dataset[String]` storing CSV rows and returns the result as a `DataFrame`.
+   *
+   * Unless the schema is specified using `schema` function, this function goes through the
+   * input once to determine the input schema.
+   *
+   * @param csvDataset input Dataset with one CSV row per record
+   * @since 2.2.0
+   */
+  def csv(csvDataset: Dataset[String]): DataFrame = {
+    val parsedOptions: CSVOptions = new CSVOptions(
+      extraOptions.toMap,
+      sparkSession.sessionState.conf.sessionLocalTimeZone)
+    val filteredLines = CSVUtils.filterCommentAndEmpty(csvDataset, parsedOptions)
+    val maybeFirstLine = filteredLines.take(1).headOption
+    if (maybeFirstLine.isEmpty) {
+      return sparkSession.emptyDataFrame
+    }
+
+    val firstLine = maybeFirstLine.get
+    val linesWithoutHeader: RDD[String] = filteredLines.rdd.mapPartitions(
+      CSVUtils.filterHeaderLine(_, firstLine, parsedOptions))
+
+    val schema = userSpecifiedSchema.getOrElse {
```

Comment:
Yes, I think we could with explicitly `wholeFile` set to false. Let me give a shot and show you if it looks okay.